### PR TITLE
add passman repo

### PR DESCRIPTION
**Pull request by enoch85**
Nextant is failing, I'm checking why... don't remove this branch please

### CODE REVIEW

**Comment by enoch85** [comment]
@morph027 I have no idea what's wrong here... I still want to add a funtion to properly test exit statues for commands but haven't come up with something good yet. Here is what I got so far but i*m not satisfied.

``` 
function test_command {
    "$@"
    local status=$?
    if [ $status -ne 0 ]; then
        echo "error with $1" >&2
    fi
    return $status
}

test_command sudo -u www-data php $NCPATH/occ app:enable nextant
test_command sudo -u www-data php $NCPATH/occ nextant:test http://127.0.0.1:8983/solr/ nextant --save
test_command sudo -u www-data php $NCPATH/occ nextant:index
``` 

Do you have something in mind for this maybe?

**Comment by morph027** [comment]
I'm using something like this:

```
check_command() {
  eval "$@"
  if [ ! $? -eq 0 ]; then
    echo "$@ failed"
    exit 1
  fi
}

check_command "sudo -u www-data uptime"
```

You need to quote the command being passed to the function.

**Comment by enoch85** [comment]
Great! But, if I want to use the exit status in an if argument to type instructions to the user..?

I did this but it only tests the first command and not all:
```
# Check that everything is OK
if  sudo -u www-data php $NCPATH/occ app:enable nextant \
&& sudo -u www-data php $NCPATH/occ nextant:test http://127.0.0.1:8983/solr/ nextant --save \
&& sudo -u www-data php $NCPATH/occ nextant:index
then
    echo "Nextant app is now installed and enabled."
    sleep 3
else
    echo "Nextant app install failed"
    sleep 3
    exit 1
fi
``` 

You had a nice for loop in the update script, haven't tested it yet, but you follow me?

**Comment by morph027** [comment]
Will think of something and let you know...

**Comment by morph027** [comment]
Loop does not help imho. We can just do this sequentially like this:
```
check_command() {
  eval "$@"
  if [ ! $? -eq 0 ]; then
    return 1
  fi
}

check_command "sudo -u www-data foo" || (echo "Error: FOO! Do foo to resolve..."; exit)
check_command "sudo -u www-data bar" || (echo "Error: BAR! Do bar to resolve..."; exit)
```

**Comment by enoch85** [comment]
Nice! But what if I want a different output every time? Hmm,maybe something standardized like:
ISSUES="https://github.com/nextcloud/vm/issues"
**Sorry but something went wrong. Please report this issue to "$ISSUES" and include the output of the error message. Thank you!** <-- in $red

Well, now it's only some polishing left. Like add some more functions to improve the code in some places:
- https://github.com/enoch85/vm/blob/patch-1/static/nextant.sh#L116 (Nextant)
- Fix the failing install_3rdparty_app function (I don't think it's the function but something fails anyway)
- Adding spinner_loading for each time the lib.sh is imported (it takes a while and the marker just blinks)
- https://github.com/enoch85/vm/blob/patch-1/nextcloud-startup-script.sh#L281 & https://github.com/enoch85/vm/blob/patch-1/nextcloud-startup-script.sh#L283(test mysql)
- Fix the colors here: https://github.com/enoch85/vm/blob/patch-1/nextcloud-startup-script.sh#L456
[ ] - Final tests
[ ] - Celebrate and donate! :D

**Comment by enoch85** [comment]
OK, tested again and the apps function works from my patch repo...

**Comment by enoch85** [comment]
**ERROR IN LETS ENCRYPT SCRIPT**
```
Certbot can obtain and install HTTPS/TLS/SSL certificates.  By default,
it will attempt to use a webserver both for obtaining and installing the
cert. 
certbot: error: unrecognized arguments: certonly --standalone --rsa-key-size 4096 --renew-by-default --agree-tos -d ssl.techandme.se
Enabling site 000-default.
To activate the new configuration, you need to run:
  service apache2 reload
It seems like no certs were generated, we do 4 more tries.Press any key to continue...
usage: 
  certbot [SUBCOMMAND] [options] [-d DOMAIN] [-d DOMAIN] ...

Certbot can obtain and install HTTPS/TLS/SSL certificates.  By default,
it will attempt to use a webserver both for obtaining and installing the
cert. 
certbot: error: unrecognized arguments:   --rsa-key-size 4096 --renew-by-default --agree-tos -d ssl.techandme.se
It seems like no certs were generated, we do 3 more tries.Press any key to continue...
usage: 
  certbot [SUBCOMMAND] [options] [-d DOMAIN] [-d DOMAIN] ...

Certbot can obtain and install HTTPS/TLS/SSL certificates.  By default,
it will attempt to use a webserver both for obtaining and installing the
cert. 
certbot: error: unrecognized arguments: certonly --webroot --w /var/www/nextcloud --rsa-key-size 4096 --renew-by-default --agree-tos -d ssl.techandme.se
It seems like no certs were generated, we do 2 more tries.Press any key to continue...
usage: 
  certbot [SUBCOMMAND] [options] [-d DOMAIN] [-d DOMAIN] ...

Certbot can obtain and install HTTPS/TLS/SSL certificates.  By default,
it will attempt to use a webserver both for obtaining and installing the
cert. 
certbot: error: unrecognized arguments: --rsa-key-size 4096 --renew-by-default --agree-tos -d ssl.techandme.se
It seems like no certs were generated, we do 1 more tries.Press any key to continue...
Sorry, last try failed as well. :/+------------------------------------------------------------------------+
| The script is located in /var/scripts/activate-ssl.sh                  |
| Please try to run it again some other time with other settings.        |
|                                                                        |
| There are different configs you can try in Let's Encrypt's user guide: |
| https://letsencrypt.readthedocs.org/en/latest/index.html               |
| Please check the guide for further information on how to enable SSL.   |
|                                                                        |
| This script is developed on GitHub, feel free to contribute:           |
| https://github.com/nextcloud/vm                                        |
|                                                                        |
| The script will now do some cleanup and revert the settings.           |
+------------------------------------------------------------------------+
Press any key to revert settings and exit... 
```

**Comment by enoch85** [comment]
**ERROR WITH APPS** Nextant and Passman
```
INFO  - 2017-04-05 20:20:15.188; org.apache.http.impl.client.DefaultRequestDirector; I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {}->http://localhost:8983: The target server failed to respond
INFO  - 2017-04-05 20:20:15.240; org.apache.http.impl.client.DefaultRequestDirector; Retrying request to {}->http://localhost:8983

ERROR: Connection refused (Connection refused)


&nextant_component;
</config>
SOLR_OPTS="$SOLR_OPTS -Dsolr.allow.unsafe.resourceloading=true"
tar (child): nextant-master-.tar.gz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: Child returned status 2
tar: Error is not recoverable: exiting now
Creating possible missing Directories
chmod Files and Directories
chown Directories
chmod/chown .htaccess
rm: cannot remove 'nextant-master-.tar.gz': No such file or directory
nextant not found

                                                                  
  [Symfony\Component\Console\Exception\CommandNotFoundException]  
  There are no commands defined in the "nextant" namespace.       
                                                                  


                                                                  
  [Symfony\Component\Console\Exception\CommandNotFoundException]  
  There are no commands defined in the "nextant" namespace.       
                                                                  

Nextant app install failed
Checking latest released version on the Passman download server and if it's possible to download...
Failed! Download is not available at the moment, try again later.
Please report this issue here: https://github.com/nextcloud/passman/issues/new
```

**Comment by enoch85** [comment]
@morph027 Can you have a look at this? ---^ I'm out of ideas. Thanks!

**Comment by enoch85** [comment]
NEXTANT_RELEASE doesen't get the version, but running the script seperatly works. So in other words ut fails, because version is missing.

**Comment by morph027** [comment]
Hm...when does this happen? So at which stage of the scripts?

Testing the parts here does not throw an error...

**Comment by morph027** [comment]
Debug mode enabled:

```
# snip #
+ wget -q -P /var/www/nextcloud/apps/ https://github.com/nextcloud/nextant/releases/download/v1.0.6/nextant-master-1.0.6.tar.gz
+ cd /var/www/nextcloud/apps/
+ tar zxf nextant-master-1.0.6.tar.gz
# snip #
```

**Comment by enoch85** [comment]
@morph027 It's when running the startup-script and choosing which scripts you want to install. Passman is fixed with this PR, but Nextant keeps failing.

**Comment by enoch85** [comment]
I think it might be occasionally as it's a curl command it might fail sometimes to getch from Github, maybe we should add a check for this, like if -z then try again in a until loop with three attempts?

**Comment by enoch85** [comment]
@morph027 Please also take a look at the Let's Encrypt script.

**Comment by enoch85** [comment]
@morph027 Shellcheck fails. It's your code. What's wrong?

**Comment by morph027** [comment]
Uh, sorry...https://github.com/koalaman/shellcheck/wiki/SC2145

make it ```eval "$*"```

**Comment by enoch85** [comment]
@morph027 Can you please fix the remaining issues [here](https://github.com/morph027/vm/pull/29#issuecomment-291978191), and also the LE function and then meege this so we can do some final tests and polishing?

Want to get this into nextcloud master asap

**Comment by morph027** [comment]
Will try my best...got some influenza right now :(

**Comment by enoch85** [comment]
Aah, hope you get better soon!

Anyway, found some more stuff. 

- This needs to be seat after the read $domain in activate-ssl.sh:
++ HTTPS_CONF=/etc/apache2/sites-available/.conf

- This ends with two // 
+++ curl -s -m 900 https://download.nextcloud.com/server/releases//

- Found the Nextant issue!
+++ curl -s https://api.github.com/repos/nextcloud/nextant/releases/latest
+++ grep tag_name
+++ cut '-d"' -f4
+++ sed -e 's|v||g'
++ NEXTANT_VERSION=
++ NT_RELEASE=nextant-master-.tar.gz

And THIS is why it's not working for me:
```
root@nextcloud:/var/scripts# curl -s https://api.github.com/repos/nextcloud/calendar/releases/latest
{
  "message": "API rate limit exceeded for MY_PUBLIC_IP. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)",
  "documentation_url": "https://developer.github.com/v3/#rate-limiting"
}
```

So in other words, it works but I'm temporarily blocked from the API. Duuh

**Comment by enoch85** [comment]
@morph027 This bugs me:

```
SOLR_OPTS= -Dsolr.allow.unsafe.resourceloading=true
Sorry but something went wrong. Please report this issue to https://github.com/nextcloud/vm/issues and include the output of the error message. Thank you!
```

When running Nextant.sh with the check_command function. Shellcheck says it's correct, but it fails in the script...

**Comment by enoch85** [comment]
This is LE with debug:

``` 
+ sleep 2
+ cat
+ LE_METHODS=()
+ LE_METHODS+=("certonly --standalone")
+ LE_METHODS+=(" ")
+ LE_METHODS+=("certonly --webroot --w $NCPATH")
+ LE_METHODS+=("--apache")
+ LE_DEFAULT_OPTIONS='--rsa-key-size 4096 --renew-by-default --agree-tos -d cloud.techandme.se'
+ NR_OF_ATTEMPTS=4
+ ATTEMPT=0
+ '[' '!' 0 -eq 4 ']'
+ case "${LE_METHODS[$ATTEMPT]}" in
+ a2dissite 000-default.conf
Site 000-default disabled.
To activate the new configuration, you need to run:
  service apache2 reload
+ sudo service apache2 stop
+ letsencrypt 'certonly --standalone' '--rsa-key-size 4096 --renew-by-default --agree-tos -d cloud.techandme.se'
usage: 
  certbot [SUBCOMMAND] [options] [-d DOMAIN] [-d DOMAIN] ...

Certbot can obtain and install HTTPS/TLS/SSL certificates.  By default,
it will attempt to use a webserver both for obtaining and installing the
cert. 
certbot: error: unrecognized arguments: certonly --standalone --rsa-key-size 4096 --renew-by-default --agree-tos -d cloud.techandme.se
```

**Comment by morph027** [comment]
```
check_command "echo \"SOLR_OPTS=\\\"\\\$SOLR_OPTS -Dsolr.allow.unsafe.resourceloading=true\\\"\" | sudo tee -a /etc/default/solr.in.sh"
```

**Comment by morph027** [comment]
The LE error is related to the ticks ```'``` around the parameters...will try to figure out why...

**Comment by enoch85** [comment]
This is the failing line:
`"${LE_METHODS[$ATTEMPT]}" "$LE_DEFAULT_OPTIONS"`
```
+ echo 'certonly --standalone' '--rsa-key-size 4096 --renew-by-default --agree-tos -d '
certonly --standalone --rsa-key-size 4096 --renew-by-default --agree-tos -d
```

**Comment by morph027** [comment]
You can try to

```
eval letsencrypt "${LE_METHODS[$ATTEMPT]}" "$LE_DEFAULT_OPTIONS"
```

**Comment by enoch85** [comment]
Got this from #bash 

`f1() { echo certbot certonly --standalone; }; f2() { echo ...; }; methods=(f1 f2); for f in "${methods[@]}"; do "$f"; done`

**Comment by enoch85** [comment]
aka make each method a function

**Comment by enoch85** [comment]
@morph027 Join me on IRC #techandme

**Comment by enoch85** [comment]
@morph027 eval worked but: http://mywiki.wooledge.org/BashFAQ/048

**Comment by enoch85** [comment]
From IRC: "[19:02] <Soliton> enoch855: in case you want to do it properly: rewrite the body of your while loop into a function and call that function several times with your different options."

**Comment by enoch85** [comment]
**READY TO MERGE TO NEXTCLOUD MASTER**
- [X] nextcloud_install_production.sh

**CLEAN UP LEFT**
- remove all `sleep` before a script is run (takes a while to load the bash anyway so no sleep is needed)
- while loop when asking if the domain is correct in activate-ssl.sh
- ALL the other issues in the comments above --^

**Comment by morph027** [comment]
Ok!

Will try if i can be creative today. Should i merge this anyway?

**Comment by enoch85** [comment]
Yes you can merge this, but don't remove the branch please.

**Comment by enoch85** [comment]
OK, **don't** merge just yet.Will clone this branch and do some fixes.

**Comment by enoch85** [comment]
OK @morph027 Please merge this so I can test it, I'm done so far with my parts.

**Comment by enoch85** [comment]
**MERGE**

**Comment by enoch85** [comment]
OK, I'll make a new branch from here. This is getting too big now. Please merge @morph027
